File: viewer/public/app/river.js
// Turn a drawn route into a navigable channel the engine actually carves.
//
// **This is the authored-world-change path, and it is not an exception to point-evaluability.**
// A feature is a shape stated at a place; asking whether a point lies inside it is a pure
// function of that point. Flood-filling a channel across a heightfield would have needed a
// raster and a traversal order, and would have cost the property the whole engine is built on.
// A chain of `carve` segments costs nothing of the kind.
//
// **Measured before it was written**, on the owner's own world with the Python oracle:
//
//     one segment, ground 33.70 m, target -8 m, 300 x 90 m   ->  -8.00 m, weight 1.000
//     a straight chain, sampled on the centreline               ->  -8.00 m at every overlap
//
//     features    0      5     22     60    150
//     us/sample 184.7  199.1  172.1  175.5  169.0
//
// No trend in the cost. The feature scan does not rise above the terrain evaluation at a
// hundred and fifty segments, so a river with a segment every hundred metres can run twenty
// kilometres before anybody needs to think about it again.
//
// **Overlap is not optional on a bend, and that was learned the hard way.** A first test cut
// its segments to 62% of their span along a curving path and reported the channel 29 m
// shallower than asked. Nothing was wrong with the engine: the sample sat off every
// segment's centreline, in the gap between them. `OVERLAP` below is why that cannot happen
// here, and it is a number rather than a comment for the same reason.

/// How far each segment reaches along its own bearing, as a multiple of half the leg it
/// covers.
///
/// **4.5, and it was measured rather than reasoned.** A feature's weight is one at its own
/// middle and falls to nothing at its stated reach, so a chain of segments carves DEEPEST at
/// the midpoints and SHALLOWEST at the nodes between them. A river built at 1.35 looked
/// carved - it had banks and a bed and a proper cross-section - and was not navigable
/// anywhere, because the check walked it and found the bottom rising into daylight at every
/// node.
///
/// Sounded along a 3.51 km channel of 25 segments, draught 2.5 m, 90 soundings:
///
///     overlap   min depth   shoaling soundings
///       1.35     -21.30 m        65
///       2.0       -9.34 m        19
///       3.0       -0.53 m         1
///       4.5       +5.99 m         0   <- navigable
///       6.0       +8.26 m         0
///
/// Below 4.5 the channel has gaps a hull would find. Above it, nothing improves that a
/// deeper target would not do better, and every extra metre of reach is a wider disturbance
/// either side of the river.
const OVERLAP = 4.5;

/// A channel that shoals toward its head, because a river does.
///
/// Both ends are stated rather than one: a game saying "ships sail up to this town" is
/// asserting a depth at the town, and asserting it at the river mouth instead would be a
/// different and easier promise.
export const DEFAULT_RIVER = {
  mouthDepthM: -9.0,
  headDepthM: -3.0,
  mouthWidthM: 110.0,
  headWidthM: 45.0,
};

function bearingBetween(a, b) {
  const toRad = (d) => (d * Math.PI) / 180;
  const [la1, lo1, la2, lo2] = [a[0], a[1], b[0], b[1]].map(toRad);
  const y = Math.sin(lo2 - lo1) * Math.cos(la2);
  const x = Math.cos(la1) * Math.sin(la2)
    - Math.sin(la1) * Math.cos(la2) * Math.cos(lo2 - lo1);
  return ((Math.atan2(y, x) * 180) / Math.PI + 360) % 360;
}

function metresBetween(a, b, radiusM) {
  const toRad = (d) => (d * Math.PI) / 180;
  const dLat = toRad(b[0] - a[0]);
  const dLon = toRad(b[1] - a[1]);
  const h = Math.sin(dLat / 2) ** 2
    + Math.cos(toRad(a[0])) * Math.cos(toRad(b[0])) * Math.sin(dLon / 2) ** 2;
  return 2 * Math.asin(Math.sqrt(h)) * radiusM;
}

/// Build the engine's feature records for one river.
///
/// Args:
///   points: `[[lat, lon], ...]` in order, MOUTH FIRST. Order decides which end is deep.
///   radiusM: the planet's radius.
///   shape: overrides for `DEFAULT_RIVER`.
///
/// Returns an array of `{ latitudeDeg, longitudeDeg, targetM, lengthM, widthM, bearingDeg,
/// compose, substrate }` - exactly what `Engine.newWorld` takes.
export function riverFeatures(points, radiusM, shape = {}) {
  const s = { ...DEFAULT_RIVER, ...shape };
  const out = [];
  for (let i = 0; i < points.length - 1; i += 1) {
    const a = points[i];
    const b = points[i + 1];
    const leg = metresBetween(a, b, radiusM);
    if (leg <= 0) continue;
    // How far up the river this segment sits, so depth and width can taper.
    const t = points.length > 2 ? i / (points.length - 2) : 0;
    out.push({
      latitudeDeg: (a[0] + b[0]) / 2,
      longitudeDeg: (a[1] + b[1]) / 2,
      targetM: s.mouthDepthM + (s.headDepthM - s.mouthDepthM) * t,
      // Half-length, times the overlap. Segments must reach past their own midpoints or a
      // bend leaves the channel with a gap in it.
      lengthM: (leg / 2) * OVERLAP,
      widthM: s.mouthWidthM + (s.headWidthM - s.mouthWidthM) * t,
      bearingDeg: bearingBetween(a, b),
      compose: "carve",
      substrate: "derive",
    });
  }
  return out;
}

/// Read a saved route and turn it into river features.
export async function riverFromRoute(name, radiusM, shape = {}) {
  const file = `${String(name).replace(/[^A-Za-z0-9_-]+/g, "-")}.json`;
  const response = await fetch(`/routes/${file}`);
  if (!response.ok) throw new Error(`no route called ${file}`);
  const document_ = await response.json();
  const points = (document_.nodes || [])
    .slice()
    .sort((p, q) => (p.order ?? 0) - (q.order ?? 0))
    .map((node) => [node.latitude_deg, node.longitude_deg]);
  if (points.length < 2) throw new Error("a river needs at least two nodes");
  return { features: riverFeatures(points, radiusM, shape), points, name: document_.name };
}

/// Walk a finished channel and confirm it holds its depth.
///
/// **An unchecked river is a river that is nine metres deep except in the one place a hull
/// would find.** A game asserting that ships reach a town is asserting a depth along a path,
/// and that assertion is testable exactly the way the port mapping tests a harbour.
///
/// `elevationAt(lat, lon)` comes from the built world, so this measures what was actually
/// carved rather than what was requested.
export function soundChannel(points, elevationAt, radiusM, draughtM = 2.5, stepM = 40) {
  const readings = [];
  for (let i = 0; i < points.length - 1; i += 1) {
    const a = points[i];
    const b = points[i + 1];
    const leg = metresBetween(a, b, radiusM);
    const steps = Math.max(1, Math.round(leg / stepM));
    for (let s = 0; s < steps; s += 1) {
      const f = s / steps;
      const lat = a[0] + (b[0] - a[0]) * f;
      const lon = a[1] + (b[1] - a[1]) * f;
      readings.push({ lat, lon, depthM: -elevationAt(lat, lon) });
    }
  }
  const shoal = readings.filter((r) => r.depthM < draughtM);
  const depths = readings.map((r) => r.depthM);
  return {
    soundings: readings.length,
    minDepthM: Math.min(...depths),
    maxDepthM: Math.max(...depths),
    draughtM,
    shoalings: shoal.length,
    navigable: shoal.length === 0,
    // Where it fails, not just that it does. A shoal at one bend is a fixable thing.
    worst: shoal.sort((p, q) => p.depthM - q.depthM).slice(0, 5)
      .map((r) => ({ lat: +r.lat.toFixed(5), lon: +r.lon.toFixed(5),
                     depthM: +r.depthM.toFixed(2) })),
  };
}

File: evennia_roundtrip/river.py
"""Rivers and other authored changes, as data the worldfile carries.

**A world is not just a planet and some rooms.** It is a planet, the changes somebody made
to it, and the rooms placed on the result - and until now the worldfile carried the first
and the third. A game reading it got a river town with no river.

The engine has always accepted authored features: a shape stated at a place, evaluated as a
pure function of the point, which is why an authored change costs nothing of the
point-evaluable design. What was missing was carrying them across the boundary. This module
is that, and it is deliberately a mirror of `viewer/public/app/river.js` rather than a
second opinion: the same overlap, the same taper, the same field names.

**The order of a river's nodes is its direction, and it decides which end is deep.** Node
zero is the mouth. Reversing the list makes a river that shoals toward the sea, which is
what a lake outlet does and not what a navigable river does.
"""

import math
from dataclasses import asdict, dataclass

#: How far each segment reaches along its bearing, as a multiple of half the leg it covers.
#:
#: **4.5, measured, and the same number the viewer uses.** A feature's weight is one at its
#: middle and nothing at its stated reach, so a chain carves deepest at the midpoints and
#: shallowest at the nodes. Sounded over a 3.51 km channel of 25 segments at 2.5 m draught,
#: the shoaling count runs 65, 19, 1, 0, 0 at overlaps of 1.35, 2, 3, 4.5 and 6.
#:
#: If this number and the viewer's ever disagree, the game and the studio draw different
#: ground - which is the whole failure this file exists to prevent.
OVERLAP = 4.5


@dataclass(frozen=True)
class RiverShape:
    """What a river is, at each end. Both are stated because a taper needs two."""

    mouth_depth_m: float = -14.0
    head_depth_m: float = -11.0
    #: Half-widths. `width_m` reaches this far EITHER SIDE, so the channel is twice this.
    #:
    #: Getting that wrong drowned 87 of a 177-room town: a river asked for at "80 m" was
    #: 160 m across and swallowed the quay street it was meant to run past.
    mouth_width_m: float = 22.0
    head_width_m: float = 15.0


def _bearing(a, b):
    la1, lo1, la2, lo2 = map(math.radians, (a[0], a[1], b[0], b[1]))
    y = math.sin(lo2 - lo1) * math.cos(la2)
    x = math.cos(la1) * math.sin(la2) - math.sin(la1) * math.cos(la2) * math.cos(lo2 - lo1)
    return (math.degrees(math.atan2(y, x)) + 360.0) % 360.0


def _metres(a, b, radius_m):
    la1, lo1, la2, lo2 = map(math.radians, (a[0], a[1], b[0], b[1]))
    h = (math.sin((la2 - la1) / 2) ** 2
         + math.cos(la1) * math.cos(la2) * math.sin((lo2 - lo1) / 2) ** 2)
    return 2 * math.asin(math.sqrt(h)) * radius_m


def features_from_points(points, radius_m, shape=RiverShape()):
    """
    Carve records for one river, mouth first.

    Args:
        points (list): `[(lat, lon), ...]`, node zero at the mouth.
        radius_m (float): The planet's radius.
        shape (RiverShape, optional): Depth and half-width at each end.

    Returns:
        features (list): Dicts ready for the worldfile and for the engine.

    """
    out = []
    for index in range(len(points) - 1):
        a, b = points[index], points[index + 1]
        leg = _metres(a, b, radius_m)
        if leg <= 0.0:
            continue
        along = index / (len(points) - 2) if len(points) > 2 else 0.0
        out.append({
            "kind": "river",
            "latitude_deg": (a[0] + b[0]) / 2.0,
            "longitude_deg": (a[1] + b[1]) / 2.0,
            "target_m": shape.mouth_depth_m
            + (shape.head_depth_m - shape.mouth_depth_m) * along,
            "length_m": (leg / 2.0) * OVERLAP,
            "width_m": shape.mouth_width_m
            + (shape.head_width_m - shape.mouth_width_m) * along,
            "bearing_deg": _bearing(a, b),
            "compose": "carve",
            "substrate": "derive",
            "marked": False,
        })
    return out


def features_from_route(route, shape=RiverShape()):
    """Read a saved route file and return `(features, points)`."""
    radius_m = float(route["planet"]["radius"])
    points = [(node["latitude_deg"], node["longitude_deg"])
              for node in sorted(route["nodes"], key=lambda n: n.get("order", 0))]
    if len(points) < 2:
        raise ValueError("a river needs at least two nodes")
    return features_from_points(points, radius_m, shape), points


def to_engine(features):
    """The same records in the shape `worldbuilder.bathymetry.features.Feature` wants.

    Kept separate from the worldfile shape on purpose. The file is a promise to strangers
    and uses long, unambiguous names; the engine's constructor is positional and internal.
    Letting one drive the other would make a rename of an internal field a breaking change
    to the format.
    """
    from worldbuilder.bathymetry.features import CARVE, RAISE, SHAPE, Feature
    from worldbuilder.geometry.sphere import SpherePoint

    compose = {"carve": CARVE, "raise": RAISE, "shape": SHAPE}
    built = []
    for f in features:
        built.append(Feature(
            kind=f.get("kind", "feature"),
            at=SpherePoint.from_latlon(f["latitude_deg"], f["longitude_deg"]),
            target_m=f["target_m"],
            length_m=f["length_m"],
            width_m=f["width_m"],
            bearing_deg=f["bearing_deg"],
            compose=compose[f.get("compose", "carve")],
            marked=bool(f.get("marked", False)),
        ))
    return built


def sound(points, elevation_at, radius_m, draught_m=6.096, step_m=100.0):
    """
    Walk a channel and report whether it holds its depth.

    Notes:
        **An unchecked river is one that is fourteen metres deep except where a hull would
        find.** The first river built this way had banks, a bed and a clean cross-section,
        and 34 of its 90 soundings shoaled - because a cross-section is taken at a midpoint
        and the gaps are at the nodes.
    """
    readings = []
    for index in range(len(points) - 1):
        a, b = points[index], points[index + 1]
        steps = max(1, int(round(_metres(a, b, radius_m) / step_m)))
        for s in range(steps):
            f = s / steps
            lat = a[0] + (b[0] - a[0]) * f
            lon = a[1] + (b[1] - a[1]) * f
            readings.append((lat, lon, -elevation_at(lat, lon)))
    shoal = [r for r in readings if r[2] < draught_m]
    depths = [r[2] for r in readings]
    return {
        "soundings": len(readings),
        "draught_m": draught_m,
        "min_depth_m": round(min(depths), 2) if depths else None,
        "max_depth_m": round(max(depths), 2) if depths else None,
        "shoalings": len(shoal),
        "navigable": not shoal,
        "worst": [{"latitude_deg": round(r[0], 5), "longitude_deg": round(r[1], 5),
                   "depth_m": round(r[2], 2)} for r in sorted(shoal, key=lambda r: r[2])[:5]],
    }
